backend/ethereum/glue/eventsub_test.go

```diff
@@ -0,0 +1,118 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package glue_test
+
+import (
+	"context"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+
+	"perun.network/go-perun/backend/ethereum/bindings/peruntoken"
+	ethchannel "perun.network/go-perun/backend/ethereum/channel"
+	"perun.network/go-perun/backend/ethereum/channel/test"
+	"perun.network/go-perun/backend/ethereum/glue"
+	"perun.network/go-perun/backend/ethereum/wallet/keystore"
+	channeltest "perun.network/go-perun/channel/test"
+	plogrus "perun.network/go-perun/log/logrus"
+	pkgtest "perun.network/go-perun/pkg/test"
+	wallettest "perun.network/go-perun/wallet/test"
+)
+
+const (
+	txGasLimit = 100000
+)
+
+func init() {
+	plogrus.Set(logrus.TraceLevel, &logrus.TextFormatter{})
+}
+
+// TestEventSub tests the `EventSub` by:
+// 1. Emit `n-x` events with x <= n
+// 2. Starting up the `EventSub`
+// 3. Emit `x` events
+// 4. Checking that `EventSub` contains all events.
+func TestEventSub(t *testing.T) {
+	n := 1000
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	rng := pkgtest.Prng(t)
+
+	// Simulated chain setup.
+	sb := test.NewSimulatedBackend()
+	ksWallet := wallettest.RandomWallet().(*keystore.Wallet)
+	account := &ksWallet.NewRandomAccount(rng).(*keystore.Account).Account
+	sb.FundAddress(ctx, account.Address)
+	cb := ethchannel.NewContractBackend(sb, keystore.NewTransactor(*ksWallet, types.NewEIP155Signer(big.NewInt(1337))))
+
+	// Setup Perun Token.
+	tokenAddr, err := ethchannel.DeployPerunToken(ctx, cb, *account, []common.Address{account.Address}, channeltest.MaxBalance)
+	require.NoError(t, err)
+	token, err := peruntoken.NewERC20(tokenAddr, cb)
+	require.NoError(t, err)
+	ct := pkgtest.NewConcurrent(t)
+
+	go ct.Stage("emitter", func(t pkgtest.ConcT) {
+		for i := 0; i < n; i++ {
+			// Send the transaction.
+			opts, err := cb.NewTransactor(ctx, txGasLimit, *account)
+			require.NoError(t, err)
+			tx, err := token.IncreaseAllowance(opts, account.Address, big.NewInt(1))
+			require.NoError(t, err)
+			// Wait for the TX to be mined.
+			_, err = cb.ConfirmTransaction(ctx, tx, *account)
+			require.NoError(t, err)
+		}
+	})
+	// Random wait to get some events before and some after the event sub is setup.
+	time.Sleep(time.Duration(rng.Int31n(500)+100) * time.Millisecond)
```

Comment:
Do we know how many events will be emitted on average before the subscription is setup? I think it would be better to deterministically ensure that at least a small amount of events have been emitted before the setup.

Reply:
I added a sync channel which allows the reader to wait until `n/4` TX were sent.